### DISCUSSION TITLE
How do I create a new_view inside a component in GPUI?

### DISCUSSION THREAD

**Discussion by ted-marozzi** [zed-industries/zed#7120] [Feature Requests]
To better explain my question I have created a minimum repro of the problem [here](https://github.com/ted-marozzi/ted-itor/blob/why-button-no-work/src/editor.rs).

The question is why when I press the button does the counter never increment? I have only ever gotten this to work but hoisting the cx.new_view() all the way to the top level build root view where the code is only run once per new window, but surely we can create new view down the tree without needing to pass them all the way along? If anyone has worked this out it would be greatly appreciated!

**Comment by mikayla-maki** [comment]
On each frame you're constructing a _new_ Counter, so it never has a chance to increment before it's dropped. Try storing the counter in the Editor :)

**Reply by ted-marozzi** [reply]
Thanks for the reply, the issue is I can't store it in the editor as that is constructed at the top level root view, which is what I am trying to avoid, I wan't to have a component that creates and manages its own state, without having to create the view at the very top level of the application.

I modified the [code](https://github.com/ted-marozzi/ted-itor/blob/why-button-no-work/src/editor.rs) here to try and store the counter in a CounterButton struct, but still the code doesn't work.

**Reply by mikayla-maki** [reply]
The problem is that you have to put the state _somewhere_ that survives across frames. So you either need to store it in the editor somewhere and hand it into the component, or you need to put it somewhere else in your application (e.g. a global, or a different view or model) that can get the state to the component. Components by their nature are stateless. We have plans to add a version that can store and return state, but those haven't shipped yet.

**Reply by ted-marozzi** [reply]
> you need to put it somewhere else in your application (e.g. a global, or a different view or model)

Okay this sounds like what I want, I want to create a view (but not the root view) that that holds the state for my counter.

Again its seems to not work as I intend: https://github.com/ted-marozzi/ted-itor/blob/why-button-no-work/src/editor.rs


All I am trying to achieve is a Component or View which displays a count, has a button to increment a count, that is wrapped up in a nice Component or View that can be called without creating the view at the top level and without storing global state. Is that currently possible? It must be if all of Zed was built on this framework right?

**Reply by mikayla-maki** [reply]
It is not currently possible. Everything you see in Zed is in a `View` whose pointer is held by another `View`, which is itself held by another `View` and so on until you get to our `Workspace` root view, which holds a `Vec<Box<dyn ItemHandle>>` to everything in the workspace from the Terminal to the Editor. `ItemHandle` itself is just a trait that the `Workspace` implements for the`View<T>` struct, if T implements `Item` (our trait for getting things like the contents of a tab from things in Zed). 

So, you will have to put the pointers somewhere in your root view or in global state and you will probably have to use something like this `Box<dyn Handle>` trick, or just use an AnyView, to get dynamic typing in Rust.

**Reply by mikayla-maki** [reply]
This is a side effect of our choice to base our object life cycles on reference counting. If no one cares about a `View` or `Model`, then we might as well recycle that memory as aggressively as we can detect it. But how do can we detect that no one cares about it? We've decided to answer that with 'if no one has a reference to it'. Therefore, it becomes the GPUI application developer's responsibility to make sure there's a valid path of references from the window (or App, for globals) to anything that you want to stick around.

**Reply by mikayla-maki** [reply]
(Note that there is an exception to this, `Element`s are allowed to get some sneaky extra state based on their position in the rendered tree, that GPUI will carry from frame to frame as long as they don't move. However, we need a little help to detect when something has moved, which is why we require you to pass an `ElementID` before using stateful interfaces.  We plan on using this to support components that work the way you're looking for, eventually. We just haven't done it yet, and there are some open questions around how to get the state back out of these components.)

**Reply by ted-marozzi** [reply]
Thanks I think I understand, so the current solutions are that do what I want seem to be:
* Store views, inside of views inside of view, all of the views must be created up front in the creation of the root application view.
* Create a custom stateful element. Looking at the Stateful<Element> struct, I couldn't see where the state could actually be stored, is it that any state in the element itself will be persisted?

**Reply by mikayla-maki** [reply]
For the stateful element, you'll have to implement two traits: `Element` and `IntoElement`. `Element` gives you access to the state in it's `request_layout()` method, and `IntoElement` gives you a chance to associate an ID with the element. Fair warning though implementing elements is a very low level API.

**Reply by ted-marozzi** [reply]
I can't thank you enough! The views inside of view technique is what I needed. I underestimated the composability of this approach as you can wrap most on the complexity up in a constructor function that takes a context and hence its easy to create multiple of the same component.

<img width="455" alt="Screenshot 2024-01-31 at 9 28 34 pm" src="https://github.com/zed-industries/zed/assets/38032037/1a96a9b9-a13d-4534-81d0-428372d13dae">
